File: tools/simulate_triple_roof_cap.py
"""
Simulation hors Unity du cap triple faîtage (plan XZ, z = deuxième coordonnée).
Exécute: python tools/simulate_triple_roof_cap.py
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Tuple

Vec2 = Tuple[float, float]


def sub(a: Vec2, b: Vec2) -> Vec2:
    return (a[0] - b[0], a[1] - b[1])


def dot(a: Vec2, b: Vec2) -> float:
    return a[0] * b[0] + a[1] * b[1]


def lensq(v: Vec2) -> float:
    return dot(v, v)


def project_segment(p: Vec2, a: Vec2, b: Vec2) -> Vec2:
    ab = sub(b, a)
    ab2 = lensq(ab)
    if ab2 < 1e-14:
        return a
    t = max(0.0, min(1.0, dot(sub(p, a), ab) / ab2))
    return (a[0] + ab[0] * t, a[1] + ab[1] * t)


def closest_broken(p: Vec2, a0: Vec2, a1: Vec2, a2: Vec2) -> Vec2:
    q_ab = project_segment(p, a0, a1)
    q_bc = project_segment(p, a1, a2)
    d_ab = lensq(sub(p, q_ab))
    d_bc = lensq(sub(p, q_bc))
    return q_ab if d_ab <= d_bc else q_bc


def tri_signed_area_xz(a: Vec2, b: Vec2, c: Vec2) -> float:
    """Demi produit vectoriel (orientation dans le plan x-z)."""
    return 0.5 * (
        (b[0] - a[0]) * (c[1] - a[1]) - (c[0] - a[0]) * (b[1] - a[1])
    )


def order_triple_along_anchors(centroid: Vec2, anchor: Vec2, anchor2: Vec2) -> Tuple[Vec2, Vec2, Vec2]:
    """Même logique que OrderTripleSummitAlongAnchorLineXZ (projection sur droite des deux ancres)."""
    e = sub(anchor2, anchor)
    el2 = lensq(e)
    eps = 1e-10
    items = []
    if el2 >= eps:
        ln = math.sqrt(el2)
        eu = (e[0] / ln, e[1] / ln)
        k_c = dot(sub(centroid, anchor), eu)
        k_a = 0.0
        k_b = ln
        items = [(k_c, centroid, 0), (k_a, anchor, 1), (k_b, anchor2, 2)]
    else:
        mid = ((anchor[0] + anchor2[0]) * 0.5, (anchor[1] + anchor2[1]) * 0.5)
        d = sub(mid, centroid)
        if lensq(d) < eps:
            d = sub(anchor2, centroid)
        if lensq(d) < eps:
            return anchor, anchor2, centroid
        mag = math.sqrt(lensq(d))
        d = (d[0] / mag, d[1] / mag)
        items = [
            (0.0, centroid, 0),
            (dot(sub(anchor, centroid), d), anchor, 1),
            (dot(sub(anchor2, centroid), d), anchor2, 2),
        ]
    items.sort(key=lambda t: (t[0], t[2]))
    return items[0][1], items[1][1], items[2][1]


def ridge_param_from_a0(p: Vec2, a0: Vec2, a1: Vec2, a2: Vec2) -> float:
    q = closest_broken(p, a0, a1, a2)
    q_ab = project_segment(q, a0, a1)
    q_bc = project_segment(q, a1, a2)
    err_ab = lensq(sub(q, q_ab))
    err_bc = lensq(sub(q, q_bc))
    len_ab = math.sqrt(max(lensq(sub(a1, a0)), 1e-14))
    len_bc = math.sqrt(max(lensq(sub(a2, a1)), 1e-14))
    len_ab_sq = lensq(sub(a1, a0))
    len_bc_sq = lensq(sub(a2, a1))
    if err_ab <= err_bc:
        t = dot(sub(q, a0), sub(a1, a0)) / len_ab_sq if len_ab_sq > 1e-14 else 0.0
        t = max(0.0, min(1.0, t))
        return len_ab * t
    t2 = dot(sub(q, a1), sub(a2, a1)) / len_bc_sq if len_bc_sq > 1e-14 else 0.0
    t2 = max(0.0, min(1.0, t2))
    return len_ab + len_bc * t2


def simulate_square_ridge_across():
    # Carré 10x10, centroïde origine
    base_corners = [(-5.0, -5.0), (5.0, -5.0), (5.0, 5.0), (-5.0, 5.0)]
    centroid = (0.0, 0.0)
    # Ancres aux deux bouts du faîtage sur l’axe X, centroïde au milieu (cas typique jaune + deux ambres)
    anchor = (-3.0, 0.0)
    anchor2 = (3.0, 0.0)
    o0, o1, o2 = order_triple_along_anchors(centroid, anchor, anchor2)
    print("Ordered ridge polyline O0,O1,O2:", o0, o1, o2)

    # Anneau haut (approximation : même que ridgeTarget à alpha=1 pour quad steps=1)
    last_ring = []
    n = 4
    for i in range(n):
        ca = base_corners[i]
        cb = base_corners[(i + 1) % n]
        pa = closest_broken(ca, o0, o1, o2)
        pb = closest_broken(cb, o0, o1, o2)
        # edgeLocalT=0 pour premier point de l’arête
        rt = pa  # simplify: vertex i at corner i uses t=0 along edge i->i+1
        last_ring.append(rt)

    # Réassign correctement comme le code C# : pour vertex i, edgeIdx=i, edgeLocalT=0 → ridgeTarget = proj(corner_i)
    last_ring = [closest_broken(base_corners[i], o0, o1, o2) for i in range(n)]

    print("Last ring XZ (should lie toward ridge):", last_ring)

    # Stitch samples: milieux des arêtes du last ring (comme les mids du cap)
    centroid_xz = centroid
    stitches = []
    for i in range(n):
        j = (i + 1) % n
        mid = ((last_ring[i][0] + last_ring[j][0]) * 0.5, (last_ring[i][1] + last_ring[j][1]) * 0.5)
        q = closest_broken(mid, o0, o1, o2)
        stitches.append(q)

    # Chaîne intérieure : sommets faîtage + stitches (sans dédoublonnage pour voir les params)
    apex_pts = [o0, o1, o2]
    chain_pts = list(apex_pts) + stitches

    def key(p):
        return ridge_param_from_a0(p, o0, o1, o2)

    chain_pts_sorted = sorted(chain_pts, key=key)
    print("Interior fan chain (sorted by arc length from O0):")
    for p in chain_pts_sorted:
        print(" ", p, "param", key(p))

    # Aires signées des triangles (centroid, Vk, Vk+1) — vue depuis +Y (normale sortante ~ signe aire)
    print("Interior fan triangle signed areas (want consistent sign for 'up' mesh):")
    for k in range(len(chain_pts_sorted) - 1):
        a = tri_signed_area_xz(centroid_xz, chain_pts_sorted[k], chain_pts_sorted[k + 1])
        print(f"  fan {k}: area={a:.6f}")

    # Problème potentiel : double couche — aire des couvertures edge cap vs fan
    print("\nEdge cap mids projection (ridge stitches):", stitches)

    # Cas L / centroïde hors segment entre ancres mais ordering met centroïde au milieu le long de la droite
    print("\n--- Cas : ancres proches d’un côté, centroïde « décalé » ---")
    centroid2 = (0.0, 1.5)
    anchor_b = (-4.0, 0.0)
    anchor2_b = (4.0, 0.0)
    p0, p1, p2 = order_triple_along_anchors(centroid2, anchor_b, anchor2_b)
    print("Ordered:", p0, p1, p2)
    # Param du centroïde sur la ligne des ancres (perpendiculairement projeté)
    kc = ridge_param_from_a0(centroid2, p0, p1, p2)
    print("Centroid ridge param:", kc)


if __name__ == "__main__":
    simulate_square_ridge_across()

File: tools/simulate_triple_roof_cap.mjs
/**
 * Simulation du cap triple faîtage (XZ). node tools/simulate_triple_roof_cap.mjs
 */

function sub(a, b) {
  return [a[0] - b[0], a[1] - b[1]];
}
function dot(a, b) {
  return a[0] * b[0] + a[1] * b[1];
}
function lensq(v) {
  return dot(v, v);
}
function projectSegment(p, a, b) {
  const ab = sub(b, a);
  const ab2 = lensq(ab);
  if (ab2 < 1e-14) return a.slice();
  const t = Math.max(0, Math.min(1, dot(sub(p, a), ab) / ab2));
  return [a[0] + ab[0] * t, a[1] + ab[1] * t];
}
function closestBroken(p, a0, a1, a2) {
  const qAb = projectSegment(p, a0, a1);
  const qBc = projectSegment(p, a1, a2);
  const dAb = lensq(sub(p, qAb));
  const dBc = lensq(sub(p, qBc));
  return dAb <= dBc ? qAb : qBc;
}
function triSignedAreaXZ(a, b, c) {
  return (
    0.5 *
    ((b[0] - a[0]) * (c[1] - a[1]) - (c[0] - a[0]) * (b[1] - a[1]))
  );
}
function orderTripleAlongAnchors(centroid, anchor, anchor2) {
  const e = sub(anchor2, anchor);
  const el2 = lensq(e);
  const eps = 1e-10;
  let items;
  if (el2 >= eps) {
    const ln = Math.sqrt(el2);
    const eu = [e[0] / ln, e[1] / ln];
    const kC = dot(sub(centroid, anchor), eu);
    items = [
      [kC, centroid],
      [0, anchor],
      [ln, anchor2],
    ];
  } else {
    const mid = [(anchor[0] + anchor2[0]) * 0.5, (anchor[1] + anchor2[1]) * 0.5];
    let d = sub(mid, centroid);
    if (lensq(d) < eps) d = sub(anchor2, centroid);
    if (lensq(d) < eps) return [anchor.slice(), anchor2.slice(), centroid.slice()];
    const mag = Math.sqrt(lensq(d));
    d = [d[0] / mag, d[1] / mag];
    items = [
      [0, centroid],
      [dot(sub(anchor, centroid), d), anchor],
      [dot(sub(anchor2, centroid), d), anchor2],
    ];
  }
  items.sort((a, b) => a[0] - b[0]);
  return [items[0][1], items[1][1], items[2][1]];
}
function ridgeParamFromA0(p, a0, a1, a2) {
  const q = closestBroken(p, a0, a1, a2);
  const qAb = projectSegment(q, a0, a1);
  const qBc = projectSegment(q, a1, a2);
  const errAb = lensq(sub(q, qAb));
  const errBc = lensq(sub(q, qBc));
  const lenAbSq = lensq(sub(a1, a0));
  const lenBcSq = lensq(sub(a2, a1));
  const lenAb = Math.sqrt(Math.max(lenAbSq, 1e-14));
  const lenBc = Math.sqrt(Math.max(lenBcSq, 1e-14));
  if (errAb <= errBc) {
    const t =
      lenAbSq > 1e-14
        ? dot(sub(q, a0), sub(a1, a0)) / lenAbSq
        : 0;
    const tc = Math.max(0, Math.min(1, t));
    return lenAb * tc;
  }
  const t2 =
    lenBcSq > 1e-14
      ? dot(sub(q, a1), sub(a2, a1)) / lenBcSq
      : 0;
  const t2c = Math.max(0, Math.min(1, t2));
  return lenAb + lenBc * t2c;
}

function lastRingWithSubdivisions(steps) {
  const baseCorners = [
    [-5, -5],
    [5, -5],
    [5, 5],
    [-5, 5],
  ];
  const centroid = [0, 0];
  const anchor = [-3, 0];
  const anchor2 = [3, 0];
  const [o0, o1, o2] = orderTripleAlongAnchors(centroid, anchor, anchor2);
  const ring = [];
  const nC = baseCorners.length;
  for (let i = 0; i < nC; i++) {
    const a = baseCorners[i];
    const b = baseCorners[(i + 1) % nC];
    const ca = baseCorners[i];
    const cb = baseCorners[(i + 1) % nC];
    for (let s = 0; s < steps; s++) {
      const t = s / steps;
      const pAlong = [
        ca[0] + (cb[0] - ca[0]) * t,
        ca[1] + (cb[1] - ca[1]) * t,
      ];
      ring.push(closestBroken(pAlong, o0, o1, o2));
    }
  }
  let area = 0;
  for (let i = 0; i < ring.length; i++) {
    const j = (i + 1) % ring.length;
    area += ring[i][0] * ring[j][1] - ring[j][0] * ring[i][1];
  }
  area *= 0.5;
  return { ring, areaAbs: Math.abs(area), o0, o1, o2 };
}

function main() {
  const baseCorners = [
    [-5, -5],
    [5, -5],
    [5, 5],
    [-5, 5],
  ];
  const centroid = [0, 0];
  const anchor = [-3, 0];
  const anchor2 = [3, 0];
  const [o0, o1, o2] = orderTripleAlongAnchors(centroid, anchor, anchor2);
  console.log("Ordered ridge O0,O1,O2:", o0, o1, o2);

  const lastRing = baseCorners.map((c) => closestBroken(c, o0, o1, o2));
  console.log("Last ring XZ:", lastRing);

  const stitches = [];
  for (let i = 0; i < 4; i++) {
    const j = (i + 1) % 4;
    const mid = [
      (lastRing[i][0] + lastRing[j][0]) * 0.5,
      (lastRing[i][1] + lastRing[j][1]) * 0.5,
    ];
    stitches.push(closestBroken(mid, o0, o1, o2));
  }
  console.log("Stitches on ridge:", stitches);

  const chain = [...[o0, o1, o2].map((x) => x.slice()), ...stitches];
  const uniq = [];
  const eps = 3e-3;
  for (const p of chain) {
    if (!uniq.some((q) => lensq(sub(p, q)) <= eps * eps)) uniq.push(p);
  }
  uniq.sort((a, b) => ridgeParamFromA0(a, o0, o1, o2) - ridgeParamFromA0(b, o0, o1, o2));
  console.log("Deduped sorted chain:", uniq);

  console.log("Interior fan signed areas (CCW in XZ => +Y normal in RH):");
  for (let k = 0; k + 1 < uniq.length; k++) {
    const ar = triSignedAreaXZ(centroid, uniq[k], uniq[k + 1]);
    console.log(`  ${k}: ${ar.toFixed(6)}`);
  }

  // Chevauchement : triangle premier fan vs stitch répété
  console.log("\n--- Problème structurel ---");
  console.log(
    "Les facettes (centroid, stitches le long du faîtage) sont COPLANAIRES Y=cste avec les facettes (bord, stitch, bord).",
    "=> doubles faces au même plan => z-fighting / profondeur bizarre si les normales diffèrent."
  );

  console.log("\n--- Avec subdivisions d’arête = 2 (même convention que Unity: t = s/steps, s = 0..steps-1) ---");
  const sub2 = lastRingWithSubdivisions(2);
  console.log("Last ring point count:", sub2.ring.length, "| polygon |area| XZ:", sub2.areaAbs.toFixed(4));
  if (sub2.areaAbs < 1e-6) {
    console.log(
      "  → Aire XZ ~ 0 attendue ici: périmètre rectangle + crête dans le plan z=0 ⇒ projections sur la crête sont colinéaires en XZ.",
      "La couverture 3D reste définie par les hauteurs Y le long de l’anneau (les pentes), pas par l’aire de cette projection."
    );
  }

  console.log("\n--- Cas centroïde décalé perpendiculairement au faîtage ---");
  const c2 = [0, 1.5];
  const [p0, p1, p2] = orderTripleAlongAnchors(c2, [-4, 0], [4, 0]);
  console.log("Ordered:", p0, p1, p2);
  console.log("Param centroid projection:", ridgeParamFromA0(c2, p0, p1, p2));
}

main();
